Spigot-Server-Patches/0333-Add-entity-count-cache.patch

```diff
@@ -0,0 +1,221 @@
+From 5b101db469123c2f92ff82fc04a8b77274296bcf Mon Sep 17 00:00:00 2001
+From: Colin Godsey <crgodsey@gmail.com>
+Date: Wed, 8 Aug 2018 10:10:06 -0600
+Subject: [PATCH] Add entity count cache
+
+
+diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
+index d6b56d685..3d973deb7 100644
+--- a/src/main/java/net/minecraft/server/MinecraftServer.java
++++ b/src/main/java/net/minecraft/server/MinecraftServer.java
+@@ -1034,6 +1034,14 @@ public abstract class MinecraftServer implements IAsyncTaskHandler, IMojangStati
+                 }
+                 // CraftBukkit end */
+ 
++                /*
++                 * Paper - clear count cache every once in a while.
++                 * Cache can never be entirely trusted because
++                 * an entity may become persistent or not
++                 * during its lifetime.
++                 */
++                if (this.ticks % 600 == 0) worldserver.clearEntityCountCache();
```

Comment:
Would like to draw attention to this too. Ultimately, we can't keep the cache up to date with 100% accuracy. An entity may become persistent or not at any time, so may not trigger the incr/decr appropriately. Given that this is just limited to the creature counts now, i think its ok if it drifts a bit (again, only under the scenario that something is adding an entity, *then* setting it persistent) and we just wipe the cache every once in a while.